File: ides/visual-studio/PeerReviewer/IpcClient.cs
using System;
using System.IO;
using System.IO.Pipes;
using System.Text;
using System.Text.Json;
using System.Web;

namespace PeerReviewer
{
    public class IpcClient
    {
        private static readonly JsonSerializerOptions JsonOptions = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            PropertyNameCaseInsensitive = true
        };

        private string PipeName
        {
            get
            {
                var username = Environment.UserName;
                return $"peer-reviewer-{username}";
            }
        }

        private string ReadToken()
        {
            var tokenPath = Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
                ".peer-reviewer",
                "session.token");
            return File.ReadAllText(tokenPath).Trim();
        }

        private string SendRequest(string method, string path, string jsonBody = null)
        {
            using (var pipe = new NamedPipeClientStream(".", PipeName, PipeDirection.InOut, PipeOptions.None))
            {
                pipe.Connect(5000);

                var token = ReadToken();
                var bodyBytes = jsonBody != null ? Encoding.UTF8.GetBytes(jsonBody) : null;

                var request = new StringBuilder();
                request.Append($"{method} {path} HTTP/1.1\r\n");
                request.Append("Host: localhost\r\n");
                request.Append($"x-peer-reviewer-token: {token}\r\n");
                if (bodyBytes != null)
                {
                    request.Append("Content-Type: application/json\r\n");
                    request.Append($"Content-Length: {bodyBytes.Length}\r\n");
                }
                request.Append("Connection: close\r\n");
                request.Append("\r\n");

                var headerBytes = Encoding.UTF8.GetBytes(request.ToString());
                pipe.Write(headerBytes, 0, headerBytes.Length);
                if (bodyBytes != null)
                {
                    pipe.Write(bodyBytes, 0, bodyBytes.Length);
                }
                pipe.Flush();

                var responseBytes = ReadAllBytes(pipe);
                var response = Encoding.UTF8.GetString(responseBytes);

                var headerEnd = response.IndexOf("\r\n\r\n", StringComparison.Ordinal);
                if (headerEnd < 0)
                    throw new InvalidOperationException("Invalid HTTP response from service");

                var headerSection = response.Substring(0, headerEnd);
                var body = response.Substring(headerEnd + 4);

                var statusLine = headerSection.Split(new[] { "\r\n" }, StringSplitOptions.None)[0];
                var statusParts = statusLine.Split(' ');
                var statusCode = statusParts.Length > 1 ? int.Parse(statusParts[1]) : 0;

                if (statusCode >= 400)
                    throw new InvalidOperationException($"peer-reviewer-service responded {statusCode}: {body}");

                return body;
            }
        }

        private byte[] ReadAllBytes(Stream stream)
        {
            using (var ms = new MemoryStream())
            {
                var buffer = new byte[4096];
                int read;
                while ((read = stream.Read(buffer, 0, buffer.Length)) > 0)
                {
                    ms.Write(buffer, 0, read);
                }
                return ms.ToArray();
            }
        }

        public string RegisterRepo(string repoPath)
        {
            var payload = JsonSerializer.Serialize(new { path = repoPath }, JsonOptions);
            var body = SendRequest("POST", "/repos", payload);
            var result = JsonSerializer.Deserialize<RepoResponse>(body, JsonOptions);
            return result?.RepoRoot ?? "";
        }

        public FindingsResponse GetAllFindings(string repoRoot)
        {
            var encoded = HttpUtility.UrlEncode(repoRoot);
            var body = SendRequest("GET", $"/findings?repo={encoded}");
            return JsonSerializer.Deserialize<FindingsResponse>(body, JsonOptions) ?? new FindingsResponse();
        }

        public void AnalyzeChanges(string repoRoot)
        {
            var encoded = HttpUtility.UrlEncode(repoRoot);
            var payload = JsonSerializer.Serialize(new { scope = "changes" }, JsonOptions);
            SendRequest("POST", $"/analyze?repo={encoded}", payload);
        }

        public void AnalyzeProject(string repoRoot)
        {
            var encoded = HttpUtility.UrlEncode(repoRoot);
            var payload = JsonSerializer.Serialize(new { scope = "project" }, JsonOptions);
            SendRequest("POST", $"/analyze?repo={encoded}", payload);
        }

        public AnalysisProgress GetAnalysisProgress(string repoRoot)
        {
            var encoded = HttpUtility.UrlEncode(repoRoot);
            var body = SendRequest("GET", $"/analyze/progress?repo={encoded}");
            return JsonSerializer.Deserialize<AnalysisProgress>(body, JsonOptions) ?? new AnalysisProgress();
        }

        public void CancelAnalysis(string repoRoot)
        {
            var encoded = HttpUtility.UrlEncode(repoRoot);
            SendRequest("POST", $"/analyze/cancel?repo={encoded}");
        }

        public PeerReviewerConfig GetConfig()
        {
            var body = SendRequest("GET", "/config");
            return JsonSerializer.Deserialize<PeerReviewerConfig>(body, JsonOptions) ?? new PeerReviewerConfig();
        }

        public PeerReviewerConfig UpdateConfig(PeerReviewerConfig config)
        {
            var payload = JsonSerializer.Serialize(config, JsonOptions);
            var body = SendRequest("PUT", "/config", payload);
            return JsonSerializer.Deserialize<PeerReviewerConfig>(body, JsonOptions) ?? new PeerReviewerConfig();
        }

        public void TestProvider(PeerReviewerConfig config)
        {
            var payload = JsonSerializer.Serialize(config, JsonOptions);
            SendRequest("POST", "/providers/test", payload);
        }

        public void DismissFinding(string findingId)
        {
            SendRequest("POST", $"/findings/{findingId}/dismiss");
        }
    }
}

File: ides/vscode/src/ipc-client.ts
import * as net from "net";
import * as os from "os";
import * as path from "path";
import * as fs from "fs";

export type Severity = "info" | "low" | "medium" | "high";

export type ProviderId = "codex" | "llama-cpp" | "claude" | "opencode" | "kiro";

export type ReviewCategory =
  | "correctness"
  | "security"
  | "penetration-testing"
  | "naming"
  | "best-practice"
  | "unintended-consequence"
  | "error-handling"
  | "performance"
  | "concurrency"
  | "resource-leak"
  | "test-coverage"
  | "api-contract"
  | "maintainability";

export interface Finding {
  id: string;
  file: string;
  startLine: number;
  endLine: number;
  severity: Severity;
  category: ReviewCategory;
  title: string;
  message: string;
  provider: ProviderId;
  dismissed: boolean;
  createdAt: string;
}

export interface PeerReviewerConfig {
  activeProvider: ProviderId;
  providers: {
    codex: { command: string; args: string[] };
    llamaCpp: { baseUrl: string };
    claude: { command: string; args: string[] };
    opencode: { command: string; args: string[] };
    kiro: { command: string; args: string[] };
  };
  systemPrompt: { mode: "default" | "append" | "replace"; text: string };
  preCommit: { blockOnFindings: boolean };
  autoAnalyse: { trigger: "disabled" | "on-save" | "periodically"; intervalMinutes: number };
  maxFilesPerRun: number | null;
  debugLogging: boolean;
}

export interface AnalysisProgress {
  total: number;
  completed: number;
  startedAt: number;
}

function resolveIpcPath(): string {
  if (process.platform === "win32") {
    return `\\\\.\\pipe\\peer-reviewer-${os.userInfo().username}`;
  }
  return path.join(os.homedir(), ".peer-reviewer", "service.sock");
}

function readToken(): string {
  const tokenPath = path.join(os.homedir(), ".peer-reviewer", "session.token");
  try {
    return fs.readFileSync(tokenPath, "utf-8").trim();
  } catch {
    return "";
  }
}

interface HttpResponse {
  statusCode: number;
  headers: Record<string, string>;
  body: string;
}

export class IpcClient {
  private ipcPath: string;
  private token: string;

  constructor() {
    this.ipcPath = resolveIpcPath();
    this.token = readToken();
  }

  refreshToken(): void {
    this.token = readToken();
  }

  async request(method: string, urlPath: string, body?: unknown): Promise<HttpResponse> {
    return new Promise((resolve, reject) => {
      const socket = net.createConnection(this.ipcPath, () => {
        const bodyStr = body !== undefined ? JSON.stringify(body) : "";
        const lines: string[] = [
          `${method} ${urlPath} HTTP/1.1`,
          `Host: localhost`,
          `x-peer-reviewer-token: ${this.token}`,
          `Content-Type: application/json`,
          `Content-Length: ${Buffer.byteLength(bodyStr)}`,
          `Connection: close`,
          ``,
          bodyStr,
        ];
        socket.write(lines.join("\r\n"));
      });

      let data = "";
      socket.on("data", (chunk) => {
        data += chunk.toString();
      });

      socket.on("end", () => {
        const headerEnd = data.indexOf("\r\n\r\n");
        if (headerEnd === -1) {
          reject(new Error("Invalid HTTP response from service"));
          return;
        }
        const headerSection = data.slice(0, headerEnd);
        const responseBody = data.slice(headerEnd + 4);

        const statusLine = headerSection.split("\r\n")[0];
        const statusMatch = statusLine.match(/HTTP\/\d\.\d\s+(\d+)/);
        const statusCode = statusMatch ? parseInt(statusMatch[1], 10) : 0;

        const headers: Record<string, string> = {};
        const headerLines = headerSection.split("\r\n").slice(1);
        for (const line of headerLines) {
          const colonIdx = line.indexOf(":");
          if (colonIdx > 0) {
            headers[line.slice(0, colonIdx).toLowerCase().trim()] = line.slice(colonIdx + 1).trim();
          }
        }

        // Handle chunked transfer encoding
        let finalBody = responseBody;
        if (headers["transfer-encoding"]?.includes("chunked")) {
          finalBody = decodeChunked(responseBody);
        }

        resolve({ statusCode, headers, body: finalBody });
      });

      socket.on("error", (err) => {
        reject(err);
      });

      socket.setTimeout(120000, () => {
        socket.destroy();
        reject(new Error("IPC request timed out"));
      });
    });
  }

  async registerRepo(repoPath: string): Promise<{ repoRoot: string }> {
    const resp = await this.request("POST", "/repos", { path: repoPath });
    if (resp.statusCode !== 200) {
      throw new Error(`registerRepo failed (${resp.statusCode}): ${resp.body}`);
    }
    return JSON.parse(resp.body);
  }

  async getAllFindings(repoRoot: string): Promise<Finding[]> {
    const resp = await this.request("GET", `/findings?repo=${encodeURIComponent(repoRoot)}`);
    if (resp.statusCode !== 200) {
      throw new Error(`getAllFindings failed (${resp.statusCode}): ${resp.body}`);
    }
    return JSON.parse(resp.body).findings;
  }

  async analyzeChanges(repoRoot: string): Promise<void> {
    const resp = await this.request("POST", `/analyze?repo=${encodeURIComponent(repoRoot)}`, { scope: "changes" });
    if (resp.statusCode !== 200) {
      throw new Error(`analyzeChanges failed (${resp.statusCode}): ${resp.body}`);
    }
  }

  async analyzeProject(repoRoot: string): Promise<void> {
    const resp = await this.request("POST", `/analyze?repo=${encodeURIComponent(repoRoot)}`, { scope: "project" });
    if (resp.statusCode !== 200) {
      throw new Error(`analyzeProject failed (${resp.statusCode}): ${resp.body}`);
    }
  }

  async getAnalysisProgress(repoRoot: string): Promise<AnalysisProgress> {
    const resp = await this.request("GET", `/analyze/progress?repo=${encodeURIComponent(repoRoot)}`);
    if (resp.statusCode !== 200) {
      throw new Error(`getAnalysisProgress failed (${resp.statusCode}): ${resp.body}`);
    }
    return JSON.parse(resp.body);
  }

  async cancelAnalysis(repoRoot: string): Promise<void> {
    const resp = await this.request("POST", `/analyze/cancel?repo=${encodeURIComponent(repoRoot)}`);
    if (resp.statusCode !== 200 && resp.statusCode !== 204) {
      throw new Error(`cancelAnalysis failed (${resp.statusCode}): ${resp.body}`);
    }
  }

  async getConfig(): Promise<PeerReviewerConfig> {
    const resp = await this.request("GET", "/config");
    if (resp.statusCode !== 200) {
      throw new Error(`getConfig failed (${resp.statusCode}): ${resp.body}`);
    }
    return JSON.parse(resp.body);
  }

  async updateConfig(config: PeerReviewerConfig): Promise<void> {
    const resp = await this.request("PUT", "/config", config);
    if (resp.statusCode !== 200 && resp.statusCode !== 204) {
      throw new Error(`updateConfig failed (${resp.statusCode}): ${resp.body}`);
    }
  }

  async testProvider(config: PeerReviewerConfig): Promise<{ ok: boolean; error?: string }> {
    const resp = await this.request("POST", "/providers/test", config);
    if (resp.statusCode === 200) {
      return JSON.parse(resp.body);
    }
    if (resp.statusCode === 502) {
      return JSON.parse(resp.body);
    }
    throw new Error(`testProvider failed (${resp.statusCode}): ${resp.body}`);
  }

  async dismissFinding(findingId: string): Promise<void> {
    const resp = await this.request("POST", `/findings/${findingId}/dismiss`);
    if (resp.statusCode !== 200 && resp.statusCode !== 204) {
      throw new Error(`dismissFinding failed (${resp.statusCode}): ${resp.body}`);
    }
  }
}

function decodeChunked(raw: string): string {
  let result = "";
  let pos = 0;
  while (pos < raw.length) {
    const lineEnd = raw.indexOf("\r\n", pos);
    if (lineEnd === -1) break;
    const sizeStr = raw.slice(pos, lineEnd).trim();
    const size = parseInt(sizeStr, 16);
    if (isNaN(size) || size === 0) break;
    const chunkStart = lineEnd + 2;
    result += raw.slice(chunkStart, chunkStart + size);
    pos = chunkStart + size + 2;
  }
  return result;
}
